apps/base-docs/docs/building-with-base/guides/onchain-generative-nfts.md

```diff
@@ -0,0 +1,876 @@
+
+Although challenging, it is possible to write a smart contract that contains all the necessary logic and data to generate json metadata and svg images, entirely onchain. It **will** be expensive to deploy, but will be as cheap as simpler contracts to mint!
+
+In this tutorial, we'll show you how to do this to create your own fully-onchain art project, similar to our sample project [TODO LINK TO EXAMPLE HERE!].
+
+---
+
+## Objectives
+
+By the end of this lesson you should be able to:
+
+- Programmatically generate and return json metadata for ERC-721 tokens
+- Deterministically construct unique svg art in a smart contract
+- Generate deterministic, pseudorandom numbers
+
+---
+
+## Prerequisites
+
+### ERC-721 Tokens
+
+This guide assumes that you are able to write, test, and deploy your own ERC-721 tokens using the Solidity programming language. If you need to learn that first, check out our content in [Base Camp] or the sections specific to [ERC-721 Tokens]!
+
+### Vector Art
+
+You'll need some familiarity with the svg art format and a basic level of ability to edit and manipulate vector art. If you don't have this, find an artist friend and collaborate!
+
+---
+
+## Creating the Art Assets
+
+To start, you'll need to build a few vector art assets and mock up an example of what your NFT might look like. Later, you'll cut these up and format them in a way that your smart contract will use to assemble unique NFTs for each minter.
+
+The mockup needs to have all of the elements you plan to have in the NFT, and you should be able to manually move things around or change colors to make it so that you can create the range of variation you want. For example:
+
+- A gradient sky in which the colors are randomized
+- One of three styles of sun, always in the same spot in the upper right corner
+- One to five clouds placed randomly in the upper half of the canvas
+- A wide mountain ridge that will always be in the middle, but slide side to side to show a different part for each NFT
+- An ocean in the foreground that is always the same
+
+If you are an artist, or are working with one, you can use the vector drawing tool of your choice to assemble your mockup. If not, you can use a number of stock art or AI tool options to assist you. If you do, make sure you understand any relevant laws or terms of service!
+
+You can also work from ours: [TODO LINK]
+
+Either way, you should end up with something similar to this:
+
+![Mockup](../../../assets/images/onchain-generative-nfts/mockup.png)
+
+## The Art of Making it Fit
+
+You'll notice that the svg is probably way to big to be placed in a smart contract. Our example is 103 KB, so we'll have to be clever to make this work.
```

Comment:
Should probably include a note about what the upper limit is for storing image/SVG data onchain.

I think you can run into the upper limit (whatever it is) fairly easily before it starts complaining about contract size.